### PR TITLE
[MHLO] unfuse batch norm ops with shape dialect, and split inference …

### PR DESCRIPTION
**Pull request by qingyunqu**
…and training pattern

1. lowering with `shape` dialect. The original lowering pattern is not a Progressive Lowering. As the `shape` dialect is more mature in MLIR, I think lowering dynamic shape pattern with `shape` dialect is more suitable.
2. split inference and training pattern for flexible integration. The original patterns have both `mhlo.batch_norm_inference` and `mhlo.batch_norm_training` lowering pattern. Sometimes we want to only lower one of them, so I split these pattern.

### CODE REVIEW

**Comment by qingyunqu** [comment]
@joker-eph Could you please review this PR? Thanks.

**Comment by gbaned** [comment]
Hi @qingyunqu Can you please resolve conflicts? Thank you!

**Comment by joker-eph** [comment]
Can you update the description with more context? Why this change?

**Comment by qingyunqu** [comment]
> Can you update the description with more context? Why this change?

Done

**Comment by qingyunqu** [comment]
@jpienaar Could you please review this PR? Thanks.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/unfuse_batch_norm.cc [hunk 1]
```diff
@@ -200,26 +196,22 @@ Value createReduce(Location loc, Value operand, Value zero,
...
-          rewriter.create<arith::MulIOp>(loc, reduceSize, reduceValues[i]);
-    }
+    Value operandShape = getShapeValue(loc, operand, rewriter);
+    Value scaleShape = getShapeValue(loc, scale, rewriter);
+    Value operandTotalSize = rewriter.create<shape::NumElementsOp>(
+        loc, rewriter.getIndexType(), operandShape);
+    Value scaleTotalSize = rewriter.create<shape::NumElementsOp>(
+        loc, rewriter.getIndexType(), scaleShape);
+    Value reduceSize = rewriter.create<shape::DivOp>(
+        loc, rewriter.getIndexType(), operandTotalSize, scaleTotalSize);
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/unfuse_batch_norm.cc:214]: Nit: I'd host out `indexType = rewriter.getIndexType()`

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/unfuse_batch_norm.cc [hunk 2]
```diff
@@ -200,26 +196,22 @@ Value createReduce(Location loc, Value operand, Value zero,
...
-// Reduce from operand to operand[feature_index]
-Value calculateReduceSize(Operation* op, Value operand,
-                          RankedTensorType operandType,
+// Reduce from operand to operand[feature_index]/scale
+Value calculateReduceSize(Operation *op, Value operand,
+                          RankedTensorType operandType, Value scale,
                           RankedTensorType scaleType, int64_t featureIndex,
-                          PatternRewriter& rewriter) {
+                          PatternRewriter &rewriter) {
   Location loc = op->getLoc();
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/unfuse_batch_norm.cc:204]: Lets use ImplicitLocOpBuilder

```c++
ImplicitLocOpBuilder b(op->getLoc(), rewriter);
```

and just do `b.create<shape::NumElementsOp>(indexType, operandShape);` below. That would IMHO make the structure clearer below.

**Review by jpienaar — APPROVED** [review]
Looks good to me, thanks!

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/unfuse_batch_norm.cc (hunk 1 above)
Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/unfuse_batch_norm.cc]: Done

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/unfuse_batch_norm.cc (hunk 2 above)
Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/unfuse_batch_norm.cc]: Done

**Review by jpienaar — APPROVED** [review]
Nice, thanks

**Comment by jpienaar** [comment]
Seems an internal failure here, I'll check

**Comment by qingyunqu** [comment]
> Seems an internal failure here, I'll check

Thanks!
